src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -122,29 +125,71 @@ public JsonConverter GetConverter(Type typeToConvert)
             }
 
             DefaultJsonTypeInfoResolver.RootDefaultInstance();
-            return GetConverterInternal(typeToConvert);
+            return GetConverterFromTypeInfo(typeToConvert);
+        }
+
+        /// <summary>
+        /// Same as GetConverter but does not root converters
+        /// </summary>
+        internal JsonConverter GetConverterFromTypeInfo(Type typeToConvert)
+        {
+            if (_cachingContext == null)
+            {
+                if (_isLockedInstance)
+                {
+                    InitializeCachingContext();
+                }
+                else
+                {
+                    // We do not want to lock options instance here but we need to return correct answer
+                    // which means we need to go through TypeInfoResolver but without caching because that's the
+                    // only place which will have correct converter for JsonSerializerContext and reflection
+                    // based resolver. It will also work correctly for combined resolvers.
+                    JsonConverter? converter = GetJsonTypeInfoFromResolver(typeToConvert)?.Converter;
+                    converter ??= GetConverterFromOptionsOrReflectionConverter(typeToConvert);
+                    return converter;
+                }
+            }
+
+            return _cachingContext.GetOrAddConverter(typeToConvert);
         }
 
-        internal JsonConverter GetConverterInternal(Type typeToConvert)
+        private JsonConverter GetConverterFromTypeInfoOrOptionsNotCached(Type typeToConvert)
         {
-            // Only cache the value once (de)serialization has occurred since new converters can be added that may change the result.
-            if (_cachingContext != null)
+            JsonConverter? converter = GetOrAddJsonTypeInfoEx(typeToConvert, configured: false, doNotThrow: true)?.Converter;
```

Comment:
If this method is "NotCached", why is it calling into `GetOrAddJsonTypeInfoEx` which does consult the cache?

Reply:
I'll add a comment on top to make that clear

Reply:
For regular scenario it's technically not needed but:
- GetConverter also needs to return answer from options and reflection even if type info is null (for compat)
- we previously cached instance which technically means we always got exactly same instance after first usage

I'm ok with removing that but we need to agree above is ok

Reply:
no, we're gonna cache everything in the hot paths; will remove converter cache